radixdlt-java/src/main/java/com/radixdlt/client/core/ledger/InMemoryAtomStore.java

```diff
@@ -53,6 +60,34 @@ private void softDeleteDependentsOf(Atom atom) {
 			});
 	}
 
+	@Override
+	public void stageParticleGroup(String uuid, ParticleGroup particleGroup) {
+		synchronized (lock) {
+			Atom stagedAtom = stagedAtoms.get(uuid);
+			if (stagedAtom == null) {
+				stagedAtom = new Atom(particleGroup, System.currentTimeMillis());
+			} else {
+				List<ParticleGroup> groups = Stream.concat(stagedAtom.particleGroups(), Stream.of(particleGroup)).collect(Collectors.toList());
+				stagedAtom = new Atom(groups, System.currentTimeMillis());
+			}
+			stagedAtoms.put(uuid, stagedAtom);
+
+			particleGroup.spunParticles().forEach(sp -> {
+				Map<Particle, Spin> index = stagedParticleIndex.getOrDefault(uuid, new HashMap<>());
+				index.put(sp.getParticle(), sp.getSpin());
+				stagedParticleIndex.put(uuid, index);
+			});
+		}
+	}
+
+	@Override
+	public List<ParticleGroup> getStagedAndClear(String uuid) {
+		synchronized (lock) {
+			final Atom atom = stagedAtoms.remove(uuid);
+			stagedParticleIndex.get(uuid).clear();
```

Comment:
danger: uuid could be null and concurrent hash map does not permit null keys, so this will NPE